wire/msgmixfactoredpoly.go

```diff
@@ -0,0 +1,259 @@
+	if err != nil {
+		return err
+	}
+
+	err = msg.writeMessageNoSignature(op, w, pver)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Hash returns the message hash calculated by WriteHash.
+//
+// Hash returns an invalid or zero hash if WriteHash has not been called yet.
+//
+// This method is not safe while concurrently calling WriteHash.
+func (msg *MsgMixFactoredPoly) Hash() chainhash.Hash {
+	return msg.hash
+}
+
+// WriteHash serializes the message to a hasher and records the sum in the
+// message's Hash field.
+//
+// The hasher's Size() must equal chainhash.HashSize, or this method will
+// panic.  This method is designed to work only with hashers returned by
+// blake256.New.
+func (msg *MsgMixFactoredPoly) WriteHash(h hash.Hash) {
+	h.Reset()
+	writeElement(h, &msg.Signature)
+	msg.writeMessageNoSignature("", h, MixVersion)
+	sum := h.Sum(msg.hash[:0])
+	if len(sum) != len(msg.hash) {
+		s := fmt.Sprintf("hasher type %T has invalid Size() for chainhash.Hash", h)
+		panic(s)
+	}
+}
+
+// writeMessageNoSignature serializes all elements of the message except for
+// the signature.  This allows code reuse between message serialization, and
+// signing and verifying these message contents.
+//
+// If w implements hash.Hash, no errors will be returned for invalid message
+// construction.
+func (msg *MsgMixFactoredPoly) writeMessageNoSignature(op string, w io.Writer, pver uint32) error {
+	_, hashing := w.(hash.Hash)
+
+	count := len(msg.Roots)
+	if !hashing && count > MaxMixMcount {
+		msg := fmt.Sprintf("too many solutions to factored polynomial [count %v, max %v]",
+			count, MaxMixMcount)
+		return messageError(op, ErrInvalidMsg, msg)
+	}
+	for _, root := range msg.Roots {
+		if !hashing && len(root) > MaxMixFieldValLen {
+			msg := "root exceeds bytes necessary to represent number in field"
+			return messageError(op, ErrInvalidMsg, msg)
+		}
+	}
+	srcount := len(msg.SeenSlotReserves)
+	if !hashing && srcount > MaxMixPeers {
+		msg := fmt.Sprintf("too many previous referenced messages [count %v, max %v]",
+			srcount, MaxMixPeers)
+		return messageError(op, ErrTooManyPrevMixMsgs, msg)
+	}
+
+	err := writeElements(w, &msg.Identity, &msg.SessionID, msg.Run)
+	if err != nil {
+		return err
+	}
+
+	err = WriteVarInt(w, pver, uint64(count))
+	if err != nil {
+		return err
+	}
+	for _, root := range msg.Roots {
+		err := WriteVarBytes(w, pver, root)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = WriteVarInt(w, pver, uint64(srcount))
+	if err != nil {
+		return err
+	}
+	for i := range msg.SeenSlotReserves {
+		err = writeElement(w, &msg.SeenSlotReserves[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteSignedData writes a tag identifying the message data, followed by all
+// message fields excluding the signature.  This is the data committed to when
+// the message is signed.
+func (msg *MsgMixFactoredPoly) WriteSignedData(h hash.Hash) {
+	WriteVarString(h, MixVersion, CmdMixFactoredPoly+"-sig")
```

Comment:
Not going to hold up this PR since `WriteVarString` doesn't actually use the parameter currently anyway, but I have noticed in all of these that they're arguably misusing the protocol version.  It is generally supposed to be passed in so that it ultimately is the negotiated protocol version.  The intention of the parameter is that some future protocol version might change the semantics and thus it would need to selectively choose based on the negotiated version.

That said, I mentioned arguably previously because I guess you could argue that given this is a signature, you might want it to be stable.  However, I don't believe there is anything long term that would need it to be stable in such a scenario.